### PR TITLE
Add support for multiple diff report formats

### PR DESCRIPTION
**Pull request by Leftwitch**
## Description
This PR aims to implement the feature described in issue #137, which allows the creation of a Markdown report containing the differences between the two compared APIs.

Currently, it displays the API differences in the Markdown report but does not include any semver warnings or errors. This omission was intentional because it wasn't clear if it was a requirement. However, we can add this functionality if needed.

Additionally, this PR lays the foundation for future report formats. It introduces an abstraction layer to the diff command, which now supports the following report formats:
- CLI (The default format, unchanged from before)
- Markdown
- JSON (Might need to be checked if we want to keep that implementation because  of #150)

When you run the diff command as you normally would, the output remains the same as before, with no changes. However, you now have the option to specify the report format and file path using the following command-line options:
- report-format
- report-file-path

For more details on how to use these options, please refer to the updated Readme or use the -h option when invoking the command.

## Type of Change

- [ X] 🚀 New feature (non-breaking change)
- [ ] 🛠️ Bug fix (non-breaking change)
- [ ] ⚠️ Breaking change (feature or bug fix which breaks existing behaviors/APIs)
- [ ] 🏗️ Code refactor
- [ ] ⚙️ Build configuration change
- [X ] 📝 Documentation
- [ ] 🧹 Chore / Housekeeping

### CODE REVIEW

**Comment by devmil** [comment]
Hi. Thank you for your contribution. I will have a look at your changes as soon as I can

**Comment by Leftwitch** [comment]
Great, I noticed that a test didn't pass. I'll address that issue. However, I hope it won't obstruct your review of the code and my approach.

Hopefully my contribution is useful for you :)

**Comment by devmil** [comment]
> Great, I noticed that a test didn't pass. I'll address that issue. However, I hope it won't obstruct your review of the code and my approach.
> 
> Hopefully my contribution is useful for you :)

The failed test seems to be a timeout issues. Maybe you had bad luck with the build runner. Maybe we need to increase the timeout

**Comment by Leftwitch** [comment]
Ah! Just saw that aswell, okay in that case I'm going to leave it as-is, and make changes after your feedback :)

File: readme/change_codes.md
Comment (devmil) [inline readme/change_codes.md:19]: is there a special reason for removal of the anchors?

**Review by devmil — COMMENTED** [review]
I added some remarks / questions.
I think we should add some (basic) unit tests for the reporter just to make sure that they work in general

File: lib/src/cli/commands/diff_command.dart
Comment (Leftwitch) [reply lib/src/cli/commands/diff_command.dart:108]: I looked into the Documentation of the args package and couldn't find anything like enum options, of course I could use something like .toString() and parse it later into the enum, but is that the direction you were aiming for or your meant something else?

File: readme/change_codes.md
Comment (Leftwitch) [reply readme/change_codes.md:19]: When trying using the anchors in the URL it didn't work for me atleast, so I decided to nest them differently so I can link from the Markdown report to the correct section, however if you know another solution to make the "old" one work we can surely use that!

**Comment by Leftwitch** [comment]
> I added some remarks / questions. I think we should add some (basic) unit tests for the reporter just to make sure that they work in general


Sounds like a good Idea, I will come up with some Basics tests and push a new commit to the branch, while doing so I'm also rebasing on main since we got a conflict ^^

File: lib/src/cli/commands/diff_command.dart
Comment (devmil) [reply lib/src/cli/commands/diff_command.dart:108]: Yes, you are right: Enums are not directly supported by the args package.
You can have a look at the DependencyCheckMode arg for mapping between enum and args:
https://github.com/bmw-tech/dart_apitool/blob/b64a5e773ae591e3ab7f54f79633bbc94d86f3f3/lib/src/cli/commands/diff_command.dart#L79

File: readme/change_codes.md
Comment (devmil) [reply readme/change_codes.md:19]: Yes, makes sense. 👍

File: lib/src/cli/commands/diff_command.dart
Comment (Leftwitch) [reply lib/src/cli/commands/diff_command.dart:108]: I looked into the already existing tests, would you prefer extending the diff_command_test to check if the command handles the arguments accordingly? or what kind of Tests you had in mind? I was also looking into the unit test, but testing just the reporter class won't do much since it's just an abstract class which does nothing by its own.


So some feedback / opinion would be appreciated so I know what exactly to implement :) 


Wish yall a nice Friday :)

Comment (devmil) [reply lib/src/cli/commands/diff_command.dart:108]: Sorry for the late reply. Today has been busy.

I meant the different concrete Reporters and on a very high level.
I think it would be best to match the existing structure in the tests and add Unit Tests for the builders in a test/unit_test/diff/report directory.

If you want you additionally could add a few high level integration tests with concrete package versions but I think the unit tests are more important.

Comment (devmil) [reply lib/src/cli/commands/diff_command.dart:108]: @Leftwitch if you need any assistance or further clarification on this then just let me know

Comment (Leftwitch) [reply lib/src/cli/commands/diff_command.dart:108]: Sorry for the late reply, yeah I was Just looking on how to write good unit Tests in Dart but did Progress slower then I personally hoped, I haven't done that in the past so It's a new skill for me to learn ^^, If you can assist that might be a good solution aswell then I can learn from the code ^^ 

Or I continue attempting and you review it when it's done?

I'm only talking about the tests here, the enum and rebase are done, I can push that if you want!

Comment (devmil) [reply lib/src/cli/commands/diff_command.dart:108]: Whatever suits you best. Either way is fine for me.

Comment (Leftwitch) [reply lib/src/cli/commands/diff_command.dart:108]: Lets go with the first option so we make some progress after all!

Comment (devmil) [reply lib/src/cli/commands/diff_command.dart]: cool. Thanks.
Yes, I will add a few tests and push them to your branch. Not sure when exactly but I think should be either today or tomorrow

Comment (devmil) [reply lib/src/cli/commands/diff_command.dart]: @Leftwitch I added examples for an integration and a unit test (Markdown reporter and console reporter).
The unit tests basically only checks that nothing crashes and the integration test checks if the file has been produced and does some probing.
I think we should avoid going too deep into the produced content as this might change too frequently.
Once we find any issues we can concentrate on them and do explicit regression tests.

Comment (Leftwitch) [reply lib/src/cli/commands/diff_command.dart]: Ah dang it, I didn't get a Notification for that one, weird, Anyways the Tests look good, I think I got the gist now of what you meant ^^

And also I agree, when we go too much Into detail the tests might become to flaky

Comment (Leftwitch) [reply lib/src/cli/commands/diff_command.dart]: Also: Since the json format also generates a file we could add a test for that case aswell I think

Comment (Leftwitch) [reply lib/src/cli/commands/diff_command.dart]: I added a Test-Case for the JSON-Report, it also attempts to parse the JSON and check for some essential properties

Comment (devmil) [reply lib/src/cli/commands/diff_command.dart]: @Leftwitch just ping me if you are done

Comment (Leftwitch) [reply lib/src/cli/commands/diff_command.dart]: I added the Json test, if you don't have anything to add I'd say the PR is done so far :)

**Review by devmil — APPROVED** [review]

**Comment by devmil** [comment]
@Leftwitch feel free to press the merge button 👍
Thanks again for your contribution!

**Comment by Leftwitch** [comment]
> @Leftwitch feel free to press the merge button 👍
> Thanks again for your contribution!

I'm glad I could help :)!
Sure more to come!

I sadly can't merge the pill request I would need write access for that ^^ Just go ahead and you merge it :)

**Pull request merged**
